File: PDSG.NET/PDSG.Core/Mapping/AttributeBuilder.cs
using PDSG.Core.Models;

namespace PDSG.Core.Mapping;

/// <summary>
/// PDSG 属性构建器 — 将回路数据转换为图块属性键值对
/// </summary>
public static class AttributeBuilder
{
    /// <summary>
    /// 为单个回路构建图块属性
    /// </summary>
    public static Dictionary<string, string> BuildAttributes(
        CircuitRecord circuit,
        BlockDefinition? blockDef)
    {
        var allAttrs = new Dictionary<string, string>
        {
            ["CIRCUIT_ID"] = circuit.CircuitId,
            ["CIRCUIT_NAME"] = circuit.CircuitName,
            ["BREAKER_MODEL"] = circuit.BreakerModel,
            ["BREAKER_POLES"] = $"{circuit.BreakerPoles}P",
            ["BREAKER_Ir"] = $"{circuit.BreakerTripCurrentA:F1}A",
            ["CABLE_TYPE"] = circuit.CableType,
            ["CABLE_SECTION"] = circuit.CableSection,
            ["LOAD_POWER"] = $"{circuit.RatedPowerKw:F2}kW",
            ["LOAD_CURRENT"] = $"{circuit.RatedCurrentA:F2}A",
            ["LOAD_TYPE"] = circuit.LoadType.ToChinese(),
            ["CT_RATIO"] = circuit.CtRatio
        };

        if (!string.IsNullOrEmpty(circuit.CabinetCode))
            allAttrs["CABINET_CODE"] = circuit.CabinetCode;
        if (!string.IsNullOrEmpty(circuit.DistributionType))
            allAttrs["DISTRIBUTION_TYPE"] = circuit.DistributionType;
        if (!string.IsNullOrEmpty(circuit.OperationMode))
            allAttrs["OPERATION_MODE"] = circuit.OperationMode;
        if (circuit.BreakerFrameCurrentA.HasValue)
            allAttrs["BREAKER_FRAME"] = $"{circuit.BreakerFrameCurrentA.Value:F0}A";
        if (!string.IsNullOrEmpty(circuit.Contactor))
            allAttrs["CONTACTOR"] = circuit.Contactor;
        if (!string.IsNullOrEmpty(circuit.ThermalRelay))
            allAttrs["THERMAL_RELAY"] = circuit.ThermalRelay;
        if (!string.IsNullOrEmpty(circuit.PowerMonitoring))
            allAttrs["POWER_MONITORING"] = circuit.PowerMonitoring;
        if (!string.IsNullOrEmpty(circuit.CableNumber))
            allAttrs["CABLE_NUMBER"] = circuit.CableNumber;

        if (circuit.LoadType == LoadType.Vfd)
        {
            allAttrs["VFD_MODEL"] = circuit.VfdModel ?? "";
            allAttrs["VFD_POWER"] = circuit.VfdPowerKw.HasValue
                ? $"{circuit.VfdPowerKw.Value:F2}kW" : "";
        }

        if (!string.IsNullOrEmpty(circuit.Remark))
            allAttrs["REMARK"] = circuit.Remark;

        if (blockDef?.Attributes.Count > 0)
        {
            return blockDef.Attributes
                .Where(tag => allAttrs.ContainsKey(tag))
                .ToDictionary(tag => tag, tag => allAttrs[tag]);
        }

        return allAttrs;
    }

    /// <summary>
    /// 批量为所有已映射回路构建属性（就地写入）
    /// </summary>
    public static void BuildAllAttributes(
        List<CircuitWithBlock> mappedCircuits,
        BlockCatalog? catalog = null)
    {
        foreach (var cwb in mappedCircuits)
        {
            var blockDef = catalog?.Find(cwb.BlockName);
            cwb.Attributes = BuildAttributes(cwb.Record, blockDef);
        }
    }
}

File: 06 excel-autocad/src/attribute_builder.py
"""PDSG 属性构建器

将回路数据转换为图块属性键值对。
"""
import logging
from typing import Dict, Optional

from .data_model import (
    BlockDefinition,
    CircuitRecord,
    CircuitWithBlock,
    LoadType,
)

logger = logging.getLogger(__name__)


def build_attributes(
    circuit: CircuitRecord,
    block_def: Optional[BlockDefinition],
) -> Dict[str, str]:
    """为单个回路构建图块属性

    Args:
        circuit: 回路记录
        block_def: 图块定义（若为 None，则构建全量属性）

    Returns:
        属性字典 {Tag: 值字符串}
    """
    # 全量属性映射（Tag -> 取值函数）
    all_attrs: Dict[str, str] = {
        "CIRCUIT_ID": circuit.circuit_id,
        "CIRCUIT_NAME": circuit.circuit_name,
        "BREAKER_MODEL": circuit.breaker_model,
        "BREAKER_POLES": f"{circuit.breaker_poles}P",
        "BREAKER_Ir": f"{circuit.breaker_trip_current_a:.1f}A",
        "CABLE_TYPE": circuit.cable_type,
        "CABLE_SECTION": circuit.cable_section,
        "LOAD_POWER": f"{circuit.rated_power_kw:.2f}kW",
        "LOAD_CURRENT": f"{circuit.rated_current_a:.2f}A",
        "LOAD_TYPE": circuit.load_type.value,
        "CT_RATIO": circuit.ct_ratio,
    }

    # --- v1.1 新增属性（可选字段） ---
    if circuit.cabinet_code:
        all_attrs["CABINET_CODE"] = circuit.cabinet_code
    if circuit.distribution_type:
        all_attrs["DISTRIBUTION_TYPE"] = circuit.distribution_type
    if circuit.operation_mode:
        all_attrs["OPERATION_MODE"] = circuit.operation_mode
    if circuit.breaker_frame_current_a is not None:
        all_attrs["BREAKER_FRAME"] = f"{circuit.breaker_frame_current_a:.0f}A"
    if circuit.contactor:
        all_attrs["CONTACTOR"] = circuit.contactor
    if circuit.thermal_relay:
        all_attrs["THERMAL_RELAY"] = circuit.thermal_relay
    if circuit.power_monitoring:
        all_attrs["POWER_MONITORING"] = circuit.power_monitoring
    if circuit.cable_number:
        all_attrs["CABLE_NUMBER"] = circuit.cable_number

    # 可选属性：变频器
    if circuit.load_type == LoadType.VFD:
        all_attrs["VFD_MODEL"] = circuit.vfd_model or ""
        if circuit.vfd_power_kw is not None:
            all_attrs["VFD_POWER"] = f"{circuit.vfd_power_kw:.2f}kW"
        else:
            all_attrs["VFD_POWER"] = ""

    # 可选属性：备注
    if circuit.remark:
        all_attrs["REMARK"] = circuit.remark

    # 若图块定义指定了属性子集，只返回子集
    if block_def and block_def.attributes:
        return {
            tag: all_attrs.get(tag, "")
            for tag in block_def.attributes
        }

    return all_attrs


def build_all_attributes(
    mapped_circuits: list,
    catalog=None,
) -> None:
    """批量为所有已映射回路构建属性（就地写入 attributes 字段）

    Args:
        mapped_circuits: List[CircuitWithBlock]
        catalog: BlockCatalog 实例（可选）
    """
    for cwb in mapped_circuits:
        block_def = None
        if catalog:
            block_def = catalog.find(cwb.block_name)
        cwb.attributes = build_attributes(cwb.record, block_def)
